arch/x86/core/ia32/x86_mmu.c

```diff
@@ -344,6 +360,169 @@ void z_x86_mmu_set_flags(struct x86_mmu_pdpt *pdpt, void *ptr, size_t size,
 	}
 }
 
+static char __aligned(MMU_PAGE_SIZE)
+	page_pool[MMU_PAGE_SIZE * CONFIG_X86_MMU_PAGE_POOL_PAGES];
+
+static char *page_pos = page_pool + sizeof(page_pool);
+
+static void *get_page(void)
+{
+	page_pos -= MMU_PAGE_SIZE;
+
+	__ASSERT(page_pos >= page_pool, "out of MMU pages\n");
+
+	return page_pos;
+}
+
+__aligned(0x20) struct x86_mmu_pdpt z_x86_kernel_pdpt;
+#ifdef CONFIG_X86_KPTI
+__aligned(0x20) struct x86_mmu_pdpt z_x86_user_pdpt;
+#endif
+
+extern char z_shared_kernel_page_start[];
+
+static inline bool is_within_system_ram(uintptr_t addr)
+{
+	return (addr >= DT_PHYS_RAM_ADDR) &&
+		(addr < (DT_PHYS_RAM_ADDR + (DT_RAM_SIZE * 1024U)));
```

Comment:
Should probably do this in 64 bit math.  Physical RAM on PAE systems can routinely blow past 4G and result in an overflow here (and thus incorrect behavior) for "correct" devicetree settings.  We don't have to support using it, but we shouldn't break if devicetree authors (or EFI/e820/multiboot loaders, etc...) don't understand our limitations.

Reply:
There is no support in Zephyr for PAE itself (and there never will be), we only use PAE tables for the NX bit and nothing else.

All kinds of things will break if we are working with memory addresses that don't fit in a uintptr_t, not just here.

I'm gonna add a BUILD_ASSERT() that RAM is <= 4GB.

Reply:
> "correct" devicetree settings.

We don't need to do this.
DTS enforces already that these values fit, if I try to kick up QEMU's RAM size to be above 4GB I get a build error:

> Error: qemu_x86.dts.pre.tmp:31.21-43 Value out of range for 32-bit array element

I don't think we need to do anything here. We are never going to support PAE in the 32-bit port, we only use PAE tables for the NX bit.